### PR TITLE
Principal surprise

### PR DESCRIPTION
**Pull request by lincmba**
Principal surprise code and tests
Poetry to manage package installation
CircleCI set up

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+authors = ["Sébastien Eustace <sebastien@eustace.io>"]
+
+[tool.poetry.dependencies]
+python = "^3.7"
+plac = "^1.1"
+geojson = "^2.5.0"
+shapely = "^1.7.0"
+folium = "^0.10.1"
+scikit-learn = "^0.22.2"
+mlrose = "^1.3.0"
```
Comment (gstuder-ona) [inline pyproject.toml]: don't need mlrose

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+[tool.poetry]
+name = "perf_tools"
+version = "0.1.0"
+description = ""
+authors = ["Sébastien Eustace <sebastien@eustace.io>"]
+
+[tool.poetry.dependencies]
+python = "^3.7"
+plac = "^1.1"
+geojson = "^2.5.0"
```
Comment (gstuder-ona) [inline pyproject.toml]: don't need geojson, shapely, folium

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+shapely = "^1.7.0"
+folium = "^0.10.1"
+scikit-learn = "^0.22.2"
+mlrose = "^1.3.0"
+jsonpickle = "^1.3"
+aiohttp-index = "^0.1"
```
Comment (gstuder-ona) [inline pyproject.toml]: dont' need aiohttp stuff

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+matplotlib = "^3.2.1"
+colorhash = "^1.0.2"
+jsonpath-ng = "^1.5.1"
+pandas = "^1.1.2"
+dask = "^2.26.0"
+numpy = "^1.19.2"
+toolz = "^0.10.0"
+tornado = "^6.0.4"
```
Comment (gstuder-ona) [inline pyproject.toml]: don't think we need tornado / fsspec?

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+geojson = "^2.5.0"
+shapely = "^1.7.0"
+folium = "^0.10.1"
+scikit-learn = "^0.22.2"
+mlrose = "^1.3.0"
+jsonpickle = "^1.3"
+aiohttp-index = "^0.1"
+matplotlib = "^3.2.1"
+colorhash = "^1.0.2"
+jsonpath-ng = "^1.5.1"
```
Comment (gstuder-ona) [inline pyproject.toml]: not sure we need jsonpath-ng either

File: pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+numpy = "^1.19.2"
+toolz = "^0.10.0"
+tornado = "^6.0.4"
+fsspec = "^0.8.2"
+
+[tool.poetry.dev-dependencies]
+black = { version = "*", allow-prereleases = true }
+pytest = "^5.3"
+flake8 = "^3.7"
+pytest-asyncio = "^0.10.0"
```
Comment (gstuder-ona) [inline pyproject.toml]: shouldn't need the pytest-asyncio but maybe useful later?

File: perf_tools/principal_surprise.py [hunk 7]
```diff
@@ -0,0 +1,835 @@
...
+
+from scipy.stats import multivariate_normal
+
+from matplotlib import pyplot
+import colorhash
+
+from collections import namedtuple
+
+
+class AbnormalTaskPerfAnalyzer:
```
Comment (gstuder-ona) [inline perf_tools/principal_surprise.py]: This is fine for first pass but suspect we probably want a second patch renaming this and importing it via __init__.py

File: .circleci/config.yml [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+          key: v1-dependencies-{{ checksum "poetry.lock" }}
+          paths:
+            - ./venv
+
+      # run tests!
+      - run:
+          name: run tests
+          command: poetry run pytest
+          environment:
+            REMOTE_DOCKER_SSH_HOST: remote-docker
```
Comment (gstuder-ona) [inline .circleci/config.yml]: shouldn't need this variable

**Review by gstuder-ona — CHANGES_REQUESTED** [review]
Can we add one more patch to kill the unneeded dependencies?  CircleCI should validate that things work fine afterwards.

File: perf_tools/principal_surprise.py (hunk 7 above)
Comment (pld) [reply perf_tools/principal_surprise.py]: I think I'd rather see this class in its own file with a file name matching the class name but as snack case

**Review by pld — COMMENTED** [review]
can we change the base folder from `perf_tools` to... `chti`? `community-health-task-investigator`

File: pyproject.toml (hunk 4 above)
Comment (lincmba) [reply pyproject.toml:24]: Installing dask using `poetry add dask` not all modules are installed, as mentioned here https://docs.dask.org/en/latest/install.html#pip . Tests fail with `ImportError: Dask dataframe requirements are not installed.` until `fsspec` and`fsspec` are installed

Comment (lincmba) [reply pyproject.toml]: Installing Dask using `poetry add dask` misses some modules https://docs.dask.org/en/latest/install.html#pip. Tests then fail with `ImportError: Dask dataframe requirements are not installed.` until `toolz` and `fsspec` are installed

File: pyproject.toml [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+[tool.poetry.dependencies]
+python = "^3.7"
+plac = "^1.1"
+scikit-learn = "^0.22.2"
+jsonpickle = "^1.3"
+matplotlib = "^3.2.1"
+colorhash = "^1.0.2"
+pandas = "^1.1.2"
+numpy = "^1.19.2"
+dask = "^2.26.0"
```
Comment (gstuder-ona) [inline pyproject.toml]: yeah the line here is:
dask = { version = "^2.9", extras = ["dataframe"]}

you just need to add the extras, like in pip. If you edit the file and then `poetry update` you should get the packages you need.

File: pyproject.toml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+name = "chti"
+version = "0.1.0"
+description = ""
+authors = ["Sébastien Eustace <sebastien@eustace.io>"]
+
+[tool.poetry.dependencies]
+python = "^3.7"
+plac = "^1.1"
+scikit-learn = "^0.22.2"
+jsonpickle = "^1.3"
```
Comment (gstuder-ona) [inline pyproject.toml]: don't need jsonpickle

File: pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+matplotlib = "^3.2.1"
+colorhash = "^1.0.2"
+pandas = "^1.1.2"
+numpy = "^1.19.2"
+dask = "^2.26.0"
+toolz = "^0.10.0"
+fsspec = "^0.8.2"
```
Comment (gstuder-ona) [inline pyproject.toml]: don't need fsspec

**Comment by moshthepitt** [comment]
:)

Do we want to set up [pre-commit](https://pre-commit.com/) with black, isort, flake8, and optionally pylint and mypy? ([xlson example](https://github.com/OpenSRP/xlson)).

I know its a nitpicky comment, but I think getting "strong/opinionated" linting and formatting assistance makes for a healthier and more standardized developer experience in the long run.

**Comment by gstuder-ona** [comment]
> :)
> 
> Do we want to set up [pre-commit](https://pre-commit.com/) with black, isort, flake8, and optionally pylint and mypy? ([xlson example](https://github.com/OpenSRP/xlson)).
> 
> I know its a nitpicky comment, but I think getting "strong/opinionated" linting and formatting assistance makes for a healthier and more standardized developer experience in the long run.

Agree - not sure if it belongs in this patch vs a second patch but that's definitely why we pull in the flake8/black tools as a dependency.  Maybe we create a ticket for this to discuss?

It can't be annoying though - I find hooking into the git commits themselves kind of the wrong place, 1) because it interferes with my local development and rebasing and 2) because it requires an extra setup step in a new repo.  I kind of like the one-liner `poetry run autostyle` as implemented here: https://github.com/onaio/canopy-factuate/blob/master/pyproject.toml#L46 

Happy to discuss more, I haven't tried pre-commit specifically: https://github.com/onaio/canopy/issues/1494

**Comment by pld** [comment]
Agree let’s add it, let’s not block this b/c of it

**Comment by gstuder-ona** [comment]
@lincmba I have a patch up here: https://github.com/onaio/community-health-task-investigator/pull/3/commits/1fecc58e2a77ce011c9379fb49d395304d0cbe2e that has what I think are the minimal pyproject.toml requirements (and some stuff that's helpful for pyenv, if you use that).  If you haven't figured it out already, the tests all run etc. with these settings.

Don't worry about merge conflicts or anything either, I can easily just rebase after your patch is done.  Renaming stuff is always finicky.

File: pyproject.toml [hunk 12]
```diff
@@ -0,0 +1,23 @@
+[tool.poetry]
+name = "chti"
+version = "0.1.0"
+description = ""
+authors = ["Sébastien Eustace <sebastien@eustace.io>"]
```
Comment (gstuder-ona) [inline pyproject.toml]: Oh - do we definitely need to change this @lincmba?  Would be embarrassing to share :-)

**Review by gstuder-ona — CHANGES_REQUESTED** [review]
Once @pld is happy I'm happy, my only nitpick is that we should definitely change the author.  devs@ona.io?

File: pyproject.toml (hunk 12 above)
Comment (lincmba) [reply pyproject.toml]: Missed that. Fixing it

**Review by pld — APPROVED** [review]

**Review by gstuder-ona — APPROVED** [review]

**Pull request merged**